### PR TITLE
Rich tagging on types and attributes

### PR DESCRIPTION
**Pull request by kdallmeyer-sr**
Add tagging to BaleenTypes and to Attributes.

BTW, this is same pattern as #121

### CODE REVIEW

**Comment by HawaiianSpork** [comment]
Interesting approach.  I was thinking rich tagging would support dynamic values.  Something like:
```kotlin

        val dogDescription = "Dog".describeAs {
            "name".type(
                type = AllowsNull(StringType()),
                required = true
            )
        }.tag(name = "dogID", withValueFromAttribute = "name")

        val data = dataOf("name" to "Fido")
        assertThat(dogDescription.validate(data).results).contains(ValidationInfo(dataTrace().tag("dogID", "Fido"), "has attribute \"name\"", data))
```
I figure this would be useful for users of baleen to add domain specific tags like identifiers.  Were you thinking we would allow users to add static tags and then the caller would have translate the static tag to content?

**Comment by kdallmeyer-sr** [comment]
The use-case I was thinking about was static tagging when you know you can group tests in a way. Then the end user can filter out test results based tags that they set.  For example, not all errors are "Critical" some might be deemed "High" so the team fixing issues can prioritize.

For example:
```kotlin
val dogDescription = "Dog".describeAs {
            "name".type(
                type = AllowsNull(StringType()),
                required = true
            ).tag("level" to "critical")

            "numLegs".type(
                type = AllowsNull(IntType()),
                required = true
            ).tag("level" to "high")
        }

val data = dataOf("name" to "Fido")
assertThat(dogDescription.validate(data).results).contains(
    ValidationInfo(dataTrace().tag("level", "critical"), "has attribute \"name\"", data)
    ValidationInfo(dataTrace().tag("level", "high"), "is missing attribute \"numLegs\"", data)
)

val errors = dogDescription.validate(data).results.filterInstanceOf<ValidationError>()
val criticalErrors = errors.filter { it.dataTrace.tags["level"] == "critical"}
```


I can certainly add dynamic tagging

**Comment by kdallmeyer-sr** [comment]
Added support for dynamic tagging from BaleenType, Validator, and AttributeDescription.

```kotlin
val productDescription = "Product".describeAs {

    // The tag() method is on StringType.
    "sku".type(StringType().tag("level", "critical").tag("sku", withValue()))

    // The tag() method is on the attribute.
    "brand_manufacturer".type(StringType(min = 1, max = 500), required = true).tag("level", "low").tag("sku", withAttributeValue("sku"))

}.tag("sku", withAttributeValue("sku"))
// Above tag is on data description.
```

**Comment by kdallmeyer-sr** [comment]
@HawaiianSpork ready for re-review.

File: baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * Wraps the Validator function with a Validator function that add tags to the datatrace
+ */
+fun Validator.tag(tag: String, value: String): Validator =
+    this.tag(tag to withConstantValue(value))
+
+/**
+ * Wraps the Validator function with a Validator function that add tags to the datatrace
+ */
+fun Validator.tag(tag: String, tagger: Tagger): Validator =
+    this.tag(tag to tagger)
```
Comment (kdallmeyer-sr) [inline baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt]: Whoops, infinite recursion here

File: baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Given a context, return the Validation result. Useful when tagging DataDescription and wanting to use same validate method
+ * signature
+ */
+fun Tagged.validate(ctx: Context): Validation {
+    return Validation(ctx, validate(ctx.dataTrace, ctx.data).asIterable())
```
Comment (kdallmeyer-sr) [inline baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt]: Not tagging properly.

File: baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt (hunk 1 above)
Comment (kdallmeyer-sr) [reply baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt]: Actually note recursive

File: baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt (hunk 2 above)
Comment (kdallmeyer-sr) [reply baleen/src/main/kotlin/com/shoprunner/baleen/types/Tagged.kt]: Actually, its working as expected. Its tagging in when in the subsequent validate method.

File: baleen-kotlin/baleen-kotlin-generator/src/test/kotlin/com/shoprunner/baleen/kotlin/DataClassGeneratorForSimpleModelsTest.kt [hunk 3]
```diff
@@ -895,4 +896,45 @@ internal class DataClassGeneratorForSimpleModelsTest {
...
+
+        """.trimIndent()
+
+        val dataClassSpecs = DataClassGenerator.encode(model)
+        val outputStream = StringWriter()
+        dataClassSpecs[model]!!.writeTo(outputStream)
+        val outputStr = outputStream.toString()
+
+        assertThat(dataClassSpecs).hasSize(1)
+        // assertThat(outputStr).isEqualToIgnoringWhitespace(expectedDataClassStr)
```
Comment (kprokopczyk) [inline baleen-kotlin/baleen-kotlin-generator/src/test/kotlin/com/shoprunner/baleen/kotlin/DataClassGeneratorForSimpleModelsTest.kt]: Do you want to uncomment or delete this line?

Comment (kdallmeyer-sr) [reply baleen-kotlin/baleen-kotlin-generator/src/test/kotlin/com/shoprunner/baleen/kotlin/DataClassGeneratorForSimpleModelsTest.kt]: Its the same as the line below. I'll delete.

Comment (kdallmeyer-sr) [reply baleen-kotlin/baleen-kotlin-generator/src/test/kotlin/com/shoprunner/baleen/kotlin/DataClassGeneratorForSimpleModelsTest.kt]: Done

**Pull request merged**